### PR TITLE
[SYCLomatic #1542] Add test for `dpct_helper_sycl_compile`

### PR DESCRIPTION
**Pull request by the-slow-one**
`dpct_helper_sycl_compile` is a migration API for CMake `cuda_compile`. This test verfies the API's 
 correctness. See _dpct.cmake_.

We are disabling the test for Windows version because of bug in a DPC++ Runtime.

### CODE REVIEW

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# ===----------------------------------------------------------------------===//
+
+cmake_minimum_required(VERSION 3.20)
+
+# project(cmake_ptx NONE) set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl")
+# find_program( dpct_bin_path NAMES dpct PATHS)
+# get_filename_component(bin_path_of_dpct ${dpct_bin_path} DIRECTORY)
+# set(dpct_cmake_file_path "${bin_path_of_dpct}/../cmake/dpct.cmake")
+# include(${dpct_cmake_file_path})
+
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# Use dpct.cmake
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  execute_process(COMMAND realpath ${DPCT_BIN_PATH}
+                  OUTPUT_VARIABLE DPCT_BIN_PATH)
+endif()
+
+get_filename_component(DPCT_BIN_DIR ${DPCT_BIN_PATH} DIRECTORY)
+set(DPCT_CMAKE_PATH "${DPCT_BIN_DIR}/../cmake/dpct.cmake")
+message(STATUS "INCLUDING ${DPCT_CMAKE_PATH}")
+include(${DPCT_CMAKE_PATH})
+
+dpct_compile(objSetOne a.dp.cpp b.dp.cpp)
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
dpct_helper_sycl_compile(objSetOne a.dp.cpp b.dp.cpp)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+set(_src_files c.dp.cpp d.dp.cpp)
+dpct_compile(objSetTwo ${_src_files})
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
dpct_helper_sycl_compile(objSetTwo ${_src_files})
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  
+  list(APPEND _allObjsLibs ${CMAKE_CURRENT_BINARY_DIR}/${libName}.so)
+  
+  add_library(Lib${_cnt} SHARED IMPORTED)
+  set_target_properties(Lib${_cnt} PROPERTIES IMPORTED_LOCATION ${CMAKE_CURRENT_BINARY_DIR}/${libName}.so)
+  
+  MATH(EXPR _cnt "${_cnt}+1")
+endforeach()
+
+add_executable(target_dpct_compile main.dp.cpp)
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
add_executable(test_dpct_helper_sycl_compile main.dp.cpp)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+target_link_libraries(target_dpct_compile PRIVATE Lib1 Lib2 Lib3 Lib4)
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
target_link_libraries(test_dpct_helper_sycl_compile PRIVATE Lib1 Lib2 Lib3 Lib4)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+add_custom_target(tdc ALL DEPENDS ${_allObjsLibs})
+add_dependencies(target_dpct_compile tdc)
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
add_dependencies(test_dpct_helper_sycl_compile tdc)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    ret = call_subprocess("cmake -G \"Unix Makefiles\" -DCMAKE_CXX_COMPILER=icpx ../")
+    if not ret:
+        print("Command 'cmake configure' failed:", test_config.command_output)
+
+    ret = call_subprocess("make")
+    if not ret:
+        print("Command 'make' failed:", test_config.command_output)
+
+    return os.path.exists("test_dpct_compile")
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py]:
```suggestion
    return os.path.exists("test_dpct_helper_sycl_compile ")
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+def build_test():
+    return True
+
+def run_test():
+    return call_subprocess("test_dpct_compile")
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py]:
```suggestion
    return call_subprocess("test_dpct_helper_sycl_compile ")
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    dpct_helper_sycl_compile.
+    """
+    # clean previous migration output
+    if os.path.exists("build"):
+        shutil.rmtree("build")
+
+    # configure and build cmake containing dpct_helper_sycl_compile
+    build_cmd = (
+        "cmake -DCMAKE_CXX_COMPILER=icpx -B build -S . "
+        "&& cmake --build build --parallel"
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py]:
```suggestion
        "&& cmake --build build"
```

Fails on Windows with that options.

**Review by tomflinda — COMMENTED** [review]
The lit test new added run failed in CI test on Windows, pls take a look.

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    # configure and build cmake containing dpct_helper_sycl_compile
+    build_cmd = (
+        "cmake -DCMAKE_CXX_COMPILER=icpx -B build -S . "
+        "&& cmake --build build"
+    )
+    ret = call_subprocess(build_cmd)
+    if not ret:
+        print(f"Command '{build_cmd}' failed:", test_config.command_output)
+        return False
+
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py]:
```suggestion
    # configure and build cmake containing dpct_helper_sycl_compile
    config_cmd = "cmake -DCMAKE_CXX_COMPILER=icpx -B build -S ."
    ret = call_subprocess(config_cmd)
    if not ret:
        print(f"Command '{config_cmd}' failed:", test_config.command_output)
        return False

    build_cmd = "cmake --build build --parallel"
    ret = call_subprocess(build_cmd)
    if not ret:
        print(f"Command '{build_cmd}' failed:", test_config.command_output)
        return False
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    if not ret:
+        print(f"Command '{config_cmd}' failed:", test_config.command_output)
+        return False
+
+    build_cmd = "cmake --build build --parallel"
+    ret = call_subprocess(build_cmd)
+    if not ret:
+        print(f"Command '{build_cmd}' failed:", test_config.command_output)
+        return False
+    # make sure the binary exists
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py:44]:
```suggestion
    
    # make sure the binary exists
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    Builds the project using cmake. The cmake list contains
+    dpct_helper_sycl_compile.
+    """
+    # clean previous migration output
+    if os.path.exists("build"):
+        shutil.rmtree("build")
+
+    # configure and build cmake containing dpct_helper_sycl_compile
+    config_cmd = "cmake -DCMAKE_CXX_COMPILER=icpx -B build -S ."
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py]:
```suggestion
    config_cmd = "cmake -G "Unix Makefiles" -DCMAKE_CXX_COMPILER=icpx -B build -S ."
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    ret = call_subprocess(config_cmd)
+    if not ret:
+        print(f"Command '{config_cmd}' failed:", test_config.command_output)
+        return False
+
+    build_cmd = "cmake --build build --parallel"
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py]:
```suggestion
    build_cmd = "make -C build"
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+  
+  list(APPEND _allObjsLibs ${CMAKE_CURRENT_BINARY_DIR}/${libName}.so)
+  
+  add_library(Lib${_cnt} SHARED IMPORTED)
+  set_target_properties(Lib${_cnt} PROPERTIES IMPORTED_LOCATION ${CMAKE_CURRENT_BINARY_DIR}/${libName}.so)
+  
+  MATH(EXPR _cnt "${_cnt}+1")
+endforeach()
+
+add_executable(test_dpct_helper_sycl_compile main.dp.cpp)
```
Comment (DeepakRajHR) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt:58]:
```suggestion
add_executable(app.run main.dp.cpp)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+target_link_libraries(test_dpct_helper_sycl_compile PRIVATE Lib1 Lib2 Lib3 Lib4)
+
+add_custom_target(tdc ALL DEPENDS ${_allObjsLibs})
+add_dependencies(test_dpct_helper_sycl_compile tdc)
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
add_executable(app.run main.dp.cpp)
target_link_libraries(app.run PRIVATE Lib1 Lib2 Lib3 Lib4)

add_custom_target(tdc ALL DEPENDS ${_allObjsLibs})
add_dependencies(app.run tdc)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,63 @@
+# ==---- CMakeLists.txt ---------------------------- cmake script file ----==//
+#
+# Copyright (C) Intel Corporation SPDX-License-Identifier: Apache-2.0 WITH
+# LLVM-exception See https://llvm.org/LICENSE.txt for license information.
+#
+# ===----------------------------------------------------------------------===//
+
+cmake_minimum_required(VERSION 3.20)
+
+project(cuda_compile_sample LANGUAGES CXX)
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
project(cuda_compile_sample NONE)
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+find_program(
+  DPCT_BIN_PATH
+  NAMES dpct
+  PATHS)
+
+if(IS_SYMLINK ${DPCT_BIN_PATH})
+  execute_process(COMMAND realpath ${DPCT_BIN_PATH}
+                  OUTPUT_VARIABLE DPCT_BIN_PATH)
+endif()
```
Comment (the-slow-one) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/CMakeLists.txt]:
```suggestion
```

File: behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    Builds the project using cmake. The cmake list contains
+    dpct_helper_sycl_compile.
+    """
+    # clean previous migration output
+    if os.path.exists("build"):
+        shutil.rmtree("build")
+
+    # configure and build cmake containing dpct_helper_sycl_compile
+    config_cmd = 'cmake -G "Unix Makefiles" -DCMAKE_CXX_COMPILER=icpx -B build -S .'
```
Comment (tomflinda) [inline behavior_tests/src/cmake_dpct_helper_sycl_compile/do_test.py:32]:
```suggestion
    config_cmd = 'cmake -G "Unix Makefiles" -DCMAKE_CXX_COMPILER=icx -B build -S .'
```

**Comment by the-slow-one** [comment]
Unable to compile an executable with linking to a shared object file created using `icpx` on *Windows*.

The following are the details.

> A case where `icpx`​ is unable to link a shared object created on Windows with an executable. 
> 1. Create the shared object for a simple a.dp.cpp (content could be any valid SYCL code)​ as following which works as expected.
> ```
> $ icpx -fsycl -shared -o sycl_a.dll -ID:/iusers/dhr/lib/dpct/bin/../include a.dp.cpp​
> ```
> 2. Create an executable while linking the shared object (sycl_a.dll​) to it, the linker fails with the following message.
> ```
> $ icpx -fsycl -o run.exe -ID:/iusers/dhr/lib/dpct/bin/../include main.dp.cpp sycl_a.dll
> sycl_a.dll : fatal error LNK1107: invalid or corrupt file: cannot read at 0x368
> icpx: error: linker command failed with exit code 1107 (use -v to see invocation)
> ```
> The result is same with Windows-specific compiler, i.e. `icx-cl`. The simple example is [here](https://intel-my.sharepoint.com/:u:/p/deepak_raj_h_r/EfD-5uOgPDBNvzBsJrSdadEBm4qp4R5gG96XiONZj4_pgg)

**Comment by tomflinda** [comment]
> `dpct_helper_sycl_compile` is a migration API for CMake `cuda_compile`. This test verfies the API's correctness. See _dpct.cmake_.
> 
> We are disabling the test for Windows version because of bug in a DPC++ Runtime. The details are here.
> 
> https://jira.devtools.intel.com/browse/CMPLRLLVM-55750

 https://jira.devtools.intel.com/browse/CMPLRLLVM-55750 is a inner URL, pls remove it from commit msg, also pls file a JIRA ticket to track the windows test disable issue and set it blocked by CMPLRLLVM-55750.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Pull request merged**